### PR TITLE
Brown-Forsythe test

### PR DESCRIPTION
**Pull request by ludwiglierhammer**
Hi guys,

I did a first try impelementation of the Brown-Forsythe test.
I looped over the dimension `realization` due to the fact that [scipy.statistics.levene](https://github.com/scipy/scipy/blob/main/scipy/stats/_morestats.py#L3174-L3176) needs one-dimensional samples.
I took the idea from [here](https://gist.github.com/cbur24/eb169254dae72a89acd66998a0fc4036) . Do you have any ideas for better performance? Maybe there is a way to get rid of the loop.

I'll continue with the test tomorrow.
Greetings,
Ludwig

### CODE REVIEW

**Comment by ludwiglierhammer** [comment]
The "problem" with dimension `realization` and `scipy.stats.levene` is the following:
`xarray.apply_ufunc` passes a sample like [[1,2,3,4,5,6]] to  `scipy.stats.levene` . The number of dimension of this sample is greater than 1. This yields to a `ValueError` (see  `scipy.stats.levene` link above). The question is how to circumvent this issue without a loop? Any suggestions?

**Comment by bzah** [comment]
Hi @ludwiglierhammer you may try to use [xarray stack](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.stack.html?highlight=stack#xarray.DataArray.stack) (and unstack) to get rid of the for loop.

The idea would be to stack `time` and `realization` into, for example, a "time_x_realisation" dimension, then call the scipy function on this new dimension and finally unstack back into `time` and `realization`.

I'm not sure how it compare in term of performance to a for loop though.

File: xclim/ensembles/_robustness.py [hunk 1]
```diff
@@ -242,7 +245,36 @@ def change_significance(
...
+                output_core_dims=[[]],
+                exclude_dims={"time"},
+                vectorize=True,
+                dask="parallelized",
+                output_dtypes=[float],
+            )
+            pvals_ = pvals_.assign_coords({"realization": realization})
+            pvals_ = pvals_.expand_dims("realization")
+            pvals += [pvals_]
+        pvals = xr.concat(pvals, dim="realization")
```
Comment (aulemahal) [inline xclim/ensembles/_robustness.py:275]:
```suggestion
        pvals = xr.apply_ufunc(
            lambda f, r: spstats.levene(f, r, center="median")[1],
            fut,
            ref,
            input_core_dims=[["time"], ["time"]],
            output_core_dims=[[]],
            exclude_dims={"time"},
            vectorize=True,
            dask="parallelized",
            output_dtypes=[float],
        )
```
Wouldn't that work? 
From xarray's doc of `apply_ufunc`:

> vectorize ([bool](https://docs.python.org/3/library/functions.html#bool), optional) – If True, then assume func only takes arrays defined over core dimensions as input and vectorize it automatically

Comment (ludwiglierhammer) [reply xclim/ensembles/_robustness.py:275]: @aulemahal: perfect, this works! Then we can change `input_core_dims` and `output_core_dims` for all methods  ('manwhitney-utest', 'ttest' and 'welch-ttest') and thus we can get rid of `realization` in the `apply_ufunc` calls. I'll push that.

Comment (ludwiglierhammer) [reply xclim/ensembles/_robustness.py]: Now "xclim/testing/tests/test_ensembles.py::test_change_significance" fails. What exactly is the difference between:
```
input_core_dims=[[realization, "time"], [realization, "time"]]
output_core_dims=[[realization]]
```
and
```
input_core_dims=[["time"], ["time"]]
output_core_dims=[[]]
```
Can we apply this change to all "methods"? If no why with "brownforsythe-test".
Sorry for so many xarray-related questions :/

**Comment by Zeitsperre** [comment]
@ludwiglierhammer Pay no heed to the failing tests for now; This will be addressed in another branch today or tomorrow.

**Comment by ludwiglierhammer** [comment]
Hi guys, how should we proceed with this PR? Can I contribute anythin else?

File: xclim/ensembles/_robustness.py [hunk 2]
```diff
@@ -101,6 +101,8 @@ def change_significance(
         Two-sided T-test, without assuming equal population variance. Same significance criterion as 'ttest'.
       'mannwhitney-utest' :
         Two-sided Mann-Whiney U-test. Same significance criterion as 'ttest'.
+      'brownforsythe-test' :
+        Brown-Forsythe test with assuming skewed, non-normal distributions. Same significance criterion as 'ttest'.
```
Comment (huard) [inline xclim/ensembles/_robustness.py:105]:
```suggestion
        Brown-Forsythe test assuming skewed, non-normal distributions. Same significance criterion as 'ttest'.
```

**Review by huard — CHANGES_REQUESTED** [review]
It seems the changes generate test failures for other tests. Once this is fixed, I think the PR will be ready to be merged.

**Comment by Zeitsperre** [comment]
Hi @ludwiglierhammer, 

We're planning on preparing another release for the end of the month. Would you be able/interested in addressing the comments in time for then? It doesn't seem as though there's much left to do.

**Comment by ludwiglierhammer** [comment]
Sorry, I did some weird things. I'll try to fix this today.

**Comment by ludwiglierhammer** [comment]
Hi, is anything missing in this PR?

**Comment by Zeitsperre** [comment]
@ludwiglierhammer I don't think so. You just need to solve the merge conflict (tests were moved around this week) and I think we can move forward!

Our next release is scheduled for May 5th!

File: xclim/ensembles/_robustness.py (hunk 1 above)
Comment (Zeitsperre) [reply xclim/ensembles/_robustness.py]: I'm not sure if I understand. @aulemahal can you follow-up on this?

File: CHANGES.rst [hunk 3]
```diff
@@ -111,6 +115,7 @@ New indicators
 * New indices and indicators for determining upwelling radiation (`shortwave_upwelling_radiation_from_net_downwelling` and `longwave_upwelling_radiation_from_net_downwelling`; CF variables `rsus` and `rlus`) from net and downwelling radiation (shortwave: `rss` and `rsds`; longwave: `rls` and `rlds`). (:pull:`1271`).
 * New indice and indicator ``{snd | snw}_season_{length | start | end}`` which generalize ``snow_cover_duration`` and  ``continuous_snow_cover_{start | end}`` to allow using these functions with variable `snw` (:pull:`1275`).
 * New indice and indicator (``dryness_index``) for estimating soil humidity classifications for winegrowing regions (based on Riou et al. (1994)). (:issue:`355`, :pull:`1235`).
+* ``ensembles.change_significance`` now support Brown-Forsythe test. (:pull:`1292`).
```
Comment (Zeitsperre) [inline CHANGES.rst:118]: This needs to be placed under v0.43!

**Comment by aulemahal** [comment]
@Zeitsperre Do you have an insight on the Labeler fail?

**Comment by Zeitsperre** [comment]
It has to do with the permissions set from that workflow. Because this is coming from a fork, the permissions need to manually be set for that action (permission to set a label from a forked action). If it occurs with other forked actions, I'll patch a fix in another PR.

Feel free to ignore.

**Review by huard — APPROVED** [review]

**Comment by Zeitsperre** [comment]
@ludwiglierhammer

It looks as though there have been some dependency updates that break our tests. We're working on this ourselves and will merge the fixes to this PR when they're ready!

**Comment by ludwiglierhammer** [comment]
Many thanks for all of your help in this PR!

**Comment by Zeitsperre** [comment]
Thanks again @ludwiglierhammer !

**Pull request merged**